clang/test/3C/difftypes1a.c

```diff
@@ -1,7 +1,11 @@
-//RUN: rm -rf %t*
-//RUN: not 3c -base-dir=%S -output-dir=%t.checked %s %S/difftypes1b.c -- 2>%t.stderr
-//RUN: grep -q "merging failed" %t.stderr
+// Since the RUN commands in difftypes1a.c and difftypes1b.c process the two
+// files in different orders and the location where the error is reported
+// depends on the order, we need to use a different diagnostic verification
+// prefix (and set of corresponding comments) for each RUN command.
 
-// The desired behavior in this case is to fail, so other checks are omitted
+//RUN: rm -rf %t*
+//RUN: 3c -base-dir=%S -output-dir=%t.checked %s %S/difftypes1b.c -- -Xclang -verify=ab-expected
 
+// ab-expected-no-diagnostics
```

Comment:
I'm surprised this ordering doesn't fail. Is it because there's no merge until the second file is processed?

Reply:
As maybe you figured out, in the `3c -base-dir=%S -output-dir=%t.checked %s %S/difftypes1b.c -- -Xclang -verify=ab-expected` test, this translation unit is processed first and generates no diagnostics; then `difftypes1b.c` is processed and generates the `ab-expected-error` marked there.

I know it's a little awkward that the verification is per translation unit and thus the `ab-expected-no-diagnostics` is needed here.  If there were an easy way to do a single verification across all translation units, I'd probably do that.  But it looks like that won't be easy because `VerifyDiagnosticConsumer` requires all translation units processed together to have the same `Preprocessor` and `ASTUnit::LoadFromCompilerInvocation` creates a different `Preprocessor` each time by default.  I might be able to override the `ASTUnit`s to use a single `Preprocessor`, but that will require more code and might break other things; I can try it if you want.

Assuming we keep the verification per translation unit, can you think of a good way to document that?  I'm also open to dropping the change to the `difftypes*` tests if you find it to be harmful on net.

Reply:
I think this test is a special case, and anyway we're writing the tests. Is `expected-no-diagnostics` required? We could leave it out of more common situations.

I don't think it's worth the effort and possible confusion to use a single `Preprocessor`. We're the user. It's enough that it works and we can explain why.

Reply:
> Is `expected-no-diagnostics` required? We could leave it out of more common situations.

It's required in each translation unit that has diagnostic verification enabled but has no other `expected-*` directives (normally because the translation unit generates no diagnostics at all, though `-Xclang -verify-ignore-unexpected=LEVEL` could also come into play).  I agree that this test is an unusual case: we have two translation units and we expect a diagnostic in one and no diagnostic in the other.  Most of our diagnostic verification tests have a single translation unit.  If one or more diagnostics are expected, obviously we have to specify them.  If we're asserting the absence of diagnostics, I think the `expected-no-diagnostics` is a helpful reminder to the reader; otherwise, the only indication that we're testing anything about diagnostics would be the `-Xclang -verify` in the RUN command.

> I don't think it's worth the effort and possible confusion to use a single `Preprocessor`. We're the user. It's enough that it works and we can explain why.

Great.